src/Razor/test/VSCode.FunctionalTest/tests/TestUtil.ts

```diff
@@ -113,50 +116,112 @@ export async function waitForDocumentUpdate(
+    // So, let's wait until either all the restores complete or the timeout fires, whichever happens first.
+    const restoresPromise = Promise.all(promiseList);
+    const timeoutPromise = new Promise(r => setTimeout(r, 60000));
+    await Promise.race([restoresPromise, timeoutPromise]);
 }
 
-export async function waitForProjectReady(directory: string) {
+export async function waitForProjectsReady(...directories: string[]) {
+    console.log('Getting test projects ready');
     await removeOldProjectRazorJsons();
-    await cleanBinAndObj(directory);
+
+    if (process.env.norestore !== 'true') {
+        console.log('Cleaning Bin and Obj');
+        await cleanBinAndObj(directories);
+    }
+
+    console.log('Activating C# extension');
     await csharpExtensionReady();
+
+    console.log('Activating HTML language features extension');
     await htmlLanguageFeaturesExtensionReady();
-    await dotnetRestore(directory);
+
+    if (process.env.norestore !== 'true') {
+        console.log('Running dotnet restore on test apps');
+        await dotnetRestore(directories);
+    }
+
+    console.log('Restarting OmniSharp');
     await restartOmniSharp();
+
+    console.log('Activating Razor extension');
     await razorExtensionReady();
-    await waitForProjectConfigured(directory);
+
+    console.log('Ensuring Obj directory in test apps');
+    await waitForObjDirectories(directories);
+
+    console.log('Ensuring project.razor.json in test apps');
     await waitForProjectsConfigured();
+
+    console.log('Test projects ready');
+}
+
+export async function waitForObjDirectories(directories: string[]) {
+    for (const directory of directories) {
+        if (!fs.existsSync(directory)) {
+            throw new Error(`Project does not exist: ${directory}`);
+        }
+
+        const objDirectory = path.join(directory, 'obj');
+        await pollUntil(() => {
+            if (findInDir(objDirectory, projectConfigFile)) {
+                return true;
+            }
+
+            return false;
+        }, /* timeout */ 10000, /* pollInterval */ 250, /* suppressError */ true, `Couldn't find obj directory in ${directory}.`);
+    }
 }
 
 export async function waitForProjectsConfigured() {
-    const csProjFiles = glob.sync(`**/*.csproj`, {cwd: testAppsRoot});
+    const csProjFiles = glob.sync(`**/*.csproj`, { cwd: testAppsRoot });
     const expectedProjects = csProjFiles.length - 1;
 
     await pollUntil(() => {
-        const files = glob.sync(`**/${projectConfigFile}`, { cwd: testAppsRoot});
+        const files = glob.sync(`**/${projectConfigFile}`, { cwd: testAppsRoot });
         return files.length === expectedProjects;
-    }, /* timeout */10000, /* pollInterval */ 500, /*suppressError */ false, `Expected to have ${expectedProjects} ${projectConfigFile}'s`);
+    }, /* timeout */ 10000, /* pollInterval */ 500, /*suppressError */ true, `Expected to have ${expectedProjects} ${projectConfigFile}'s`);
```

Comment:
Suppressing here would only happen if a directory had bad permissions? Not sure why we're suppressing here

Reply:
This check seems to verify if we have `project.razor.json` in all of our test apps. But depending on which tests we run, we may only have restored a subset of the test apps. So expecting this to be true all the time doesn't work

Reply:
I can remove this entire method if you guys think it's not valuable to have.

Reply:
Ah, I see. This check was here initially to determine if O# had loaded our projects successfully which  makes me hesitate to remove it. Is there a way we can repurpose it to only target the projects we're restoring?

Reply:
We already do this in `WaitForObjDirectories` method.